drivers/sensor/mc3419/mc3419_trigger.c

```diff
@@ -0,0 +1,178 @@
+					  struct mc3419_driver_data, gpio_cb);
+
+	const struct mc3419_config *cfg = data->dev->config;
+
+	if ((pin_mask & BIT(cfg->int_gpio.pin)) == 0U) {
+		return;
+	}
+
+#if defined(CONFIG_MC3419_TRIGGER_OWN_THREAD)
+	k_sem_give(&data->trig_sem);
+#else
+	k_work_submit(&data->work);
+#endif
+}
+
+static void mc3419_process_int(const struct device *dev)
+{
+	int ret = 0;
+	const struct mc3419_config *cfg = dev->config;
+	const struct mc3419_driver_data *data = dev->data;
+	uint8_t int_source = 0;
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, MC3419_REG_INT_STATUS, &int_source);
+	if (ret < 0) {
+		goto exit;
+	}
+
+	if (int_source & MC3419_DATA_READY_MASK) {
+		if (data->handler[MC3419_TRIG_DATA_READY]) {
+			data->handler[MC3419_TRIG_DATA_READY](dev,
+				data->trigger[MC3419_TRIG_DATA_READY]);
+		}
+	}
+
+	if (int_source & MC3419_ANY_MOTION_MASK) {
+		if (data->handler[MC3419_TRIG_ANY_MOTION]) {
+			data->handler[MC3419_TRIG_ANY_MOTION](dev,
+				data->trigger[MC3419_TRIG_ANY_MOTION]);
+		}
+	}
+exit:
+	ret = i2c_reg_write_byte_dt(&cfg->i2c, MC3419_REG_INT_STATUS,
+				    MC3419_INT_CLEAR);
+	if (ret < 0) {
+		LOG_ERR("Failed to clear interrupt (%d)", ret);
+	}
+}
+
+#if defined(CONFIG_MC3419_TRIGGER_OWN_THREAD)
+static void mc3419_thread(struct mc3419_driver_data *data)
+{
+	while (1) {
+		k_sem_take(&data->trig_sem, K_FOREVER);
+		mc3419_process_int(data->dev);
+	}
+}
+#else
+static void mc3419_work_cb(struct k_work *work)
+{
+	struct mc3419_driver_data *data = CONTAINER_OF(work,
+					  struct mc3419_driver_data, work);
+
+	mc3419_process_int(data->dev);
+}
+#endif
+
+int mc3419_configure_trigger(const struct device *dev,
+			     const struct sensor_trigger *trig,
+			     sensor_trigger_handler_t handler)
+{
+	int ret = 0;
+	uint8_t buf = 0;
+	const struct mc3419_config *cfg = dev->config;
+	struct mc3419_driver_data *data = dev->data;
+
+	if (trig->type & SENSOR_TRIG_DATA_READY) {
+		data->handler[MC3419_TRIG_DATA_READY] = handler;
+		data->trigger[MC3419_TRIG_DATA_READY] = trig;
+		buf |= MC3419_DATA_READY_MASK;
+	}
+
+	if (trig->type & SENSOR_TRIG_MOTION) {
+		uint8_t int_mask = MC3419_ANY_MOTION_MASK;
+
+		buf |= MC3419_ANY_MOTION_MASK;
+		data->handler[MC3419_TRIG_ANY_MOTION] = handler;
+		data->trigger[MC3419_TRIG_ANY_MOTION] = trig;
+
+		ret = i2c_reg_update_byte_dt(&cfg->i2c, MC3419_REG_MOTION_CTRL,
+					     int_mask, handler ? int_mask : 0);
+		if (ret < 0) {
+			LOG_ERR("Failed to configure motion interrupt (%d)", ret);
+			return ret;
+		}
+	}
+	if (!(trig->type & SENSOR_TRIG_DATA_READY) &&
+	    !(trig->type & SENSOR_TRIG_MOTION)) {
+		LOG_ERR("Unsupported sensor trigger");
+		return -ENOTSUP;
+	}
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, MC3419_REG_INT_CTRL,
+				     MC3419_DATA_READY_MASK | MC3419_ANY_MOTION_MASK,
+				     buf);
+	if (ret < 0) {
+		LOG_ERR("Failed to configure interrupt (%d)", ret);
+		return ret;
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->int_gpio, GPIO_INT_EDGE_FALLING);
```

Comment:
Any specific reason for falling edge? Not all the SoC's support edge triggered interrupt in GPIO